src/client/UserPerm.h

```diff
@@ -65,8 +65,8 @@ struct UserPerm
     return *this;
   }
 
-  uid_t uid() const { return m_uid; }
-  gid_t gid() const { return m_gid; }
+  uid_t uid() const { return m_uid != (uid_t)-1 ? m_uid : ::geteuid(); }
+  gid_t gid() const { return m_gid != (gid_t)-1 ? m_gid : ::getegid(); }
```

Comment:
For the record, we should really be using the fsuid and fsgid here (at least on Linux) and not the euid/egid. In most cases, they are identical but there are cases where they can be different. Unfortunately, Linux has no getfsuid() syscall -- you have to issue setfsuid(-1) which will always fail, but will return the existing fsuid.

What we should probably do is roll our own getfsuid() routine that does that on Linux and just calls out to geteuid() on other platforms. That said, we should probably do that in the context of a different PR.